src/NHibernate/Loader/DefaultEntityAliases.cs

```diff
@@ -89,27 +100,32 @@ private string[] GetUserProvidedAlias(string propertyPath)
 			return result;
 		}
 
-		private string GetUserProvidedAlias(string propertyPath, string defaultAlias)
+		private string GetUserProvidedAlias(string propertyPath, Func<string> getDefaultAlias)
 		{
 			string[] columns = propertyPath == null ? null : GetUserProvidedAlias(propertyPath);
 			if (columns == null)
 			{
-				return defaultAlias;
+				return getDefaultAlias();
 			}
 			else
 			{
 				return columns[0];
 			}
 		}
 
+		/// <summary>
+		/// Returns aliases for subclass persister
+		/// </summary>
 		public string[][] GetSuffixedPropertyAliases(ILoadable persister)
 		{
+			if (userProvidedAliases == null)
+				return GetPropertiesAliases(persister);
+
 			int size = persister.PropertyNames.Length;
 			string[][] suffixedPropertyAliases = new string[size][];
 			for (int j = 0; j < size; j++)
 			{
-				suffixedPropertyAliases[j] = GetUserProvidedAliases(persister.PropertyNames[j], GetPropertyAliases(persister, j));
-				Intern(suffixedPropertyAliases[j]);
+				suffixedPropertyAliases[j] = GetUserProvidedAliases(persister.PropertyNames[j], () => GetPropertyAliases(persister, j));
```

Comment:
The `() => GetPropertyAliases(persister, j)` part is incorrect. When the lambda is resolved `j` would always be equal to `size`. See here: https://dotnetfiddle.net/wv8xC1

Reply:
@hazzik I don't think that your example is equal to mine. This lambda is resolved right inside cycle with closured variable. And in you examples you execute action after cycle with modified closured variable.

Reply:
Yeah, it seems that you're right.